modules/ibc-hooks/wasm_hook.go

```diff
@@ -0,0 +1,431 @@
+		return isWasmRouted, sdk.AccAddress{}, nil,
+			fmt.Errorf(types.ErrBadMetadataFormatMsg, memo, `Could not find key wasm["contract"]`)
+	}
+
+	contractAddr, err = sdk.AccAddressFromBech32(contract)
+	if err != nil {
+		return isWasmRouted, sdk.AccAddress{}, nil,
+			fmt.Errorf(types.ErrBadMetadataFormatMsg, memo, `wasm["contract"] is not a valid bech32 address`)
+	}
+
+	// The contract and the receiver should be the same for the packet to be valid
+	if contract != receiver {
+		return isWasmRouted, sdk.AccAddress{}, nil,
+			fmt.Errorf(types.ErrBadMetadataFormatMsg, memo, `wasm["contract"] should be the same as the receiver of the packet`)
+	}
+
+	// Ensure the message key is provided
+	if wasm["msg"] == nil {
+		return isWasmRouted, sdk.AccAddress{}, nil,
+			fmt.Errorf(types.ErrBadMetadataFormatMsg, memo, `Could not find key wasm["msg"]`)
+	}
+
+	// Make sure the msg key is a map. If it isn't, return an error
+	_, ok = wasm["msg"].(map[string]interface{})
+	if !ok {
+		return isWasmRouted, sdk.AccAddress{}, nil,
+			fmt.Errorf(types.ErrBadMetadataFormatMsg, memo, `wasm["msg"] is not a map object`)
+	}
+
+	// Get the message string by serializing the map
+	msgBytes, err = json.Marshal(wasm["msg"])
+	if err != nil {
+		// The tokens will be returned
+		return isWasmRouted, sdk.AccAddress{}, nil,
+			fmt.Errorf(types.ErrBadMetadataFormatMsg, memo, err.Error())
+	}
+
+	return isWasmRouted, contractAddr, msgBytes, nil
+}
+
+func (h WasmHooks) SendPacketOverride(i ICS4Middleware, ctx sdk.Context, chanCap *capabilitytypes.Capability, sourcePort string, sourceChannel string, timeoutHeight ibcclienttypes.Height, timeoutTimestamp uint64, data []byte) (sequence uint64, err error) {
+	isIcs20, icsdata := isIcs20Packet(data)
+	if !isIcs20 {
+		return i.channel.SendPacket(ctx, chanCap, sourcePort, sourceChannel, timeoutHeight, timeoutTimestamp, data) // continue
+	}
+
+	isCallbackRouted, metadata := jsonStringHasKey(icsdata.GetMemo(), types.IBCCallbackKey)
+	if !isCallbackRouted {
+		return i.channel.SendPacket(ctx, chanCap, sourcePort, sourceChannel, timeoutHeight, timeoutTimestamp, data) // continue
+	}
+
+	// We remove the callback metadata from the memo as it has already been processed.
+
+	// If the only available key in the memo is the callback, we should remove the memo
+	// from the data completely so the packet is sent without it.
+	// This way receiver chains that are on old versions of IBC will be able to process the packet
+
+	callbackRaw := metadata[types.IBCCallbackKey] // This will be used later.
+	delete(metadata, types.IBCCallbackKey)
+	bzMetadata, err := json.Marshal(metadata)
+	if err != nil {
+		return 0, errors.Wrap(err, "Send packet with callback error")
+	}
+	stringMetadata := string(bzMetadata)
+	if stringMetadata == "{}" {
+		icsdata.Memo = ""
+	} else {
+		icsdata.Memo = stringMetadata
+	}
+	dataBytes, err := json.Marshal(icsdata)
+	if err != nil {
+		return 0, errors.Wrap(err, "Send packet with callback error")
```

Comment:
while we're here, we can probably change this error and the one above to be a bit more explicit (and different from each other). This will make the change state breaking, but we're already changing the ibc version, so that shouldn't matter here (it would if we backport this to IBCv4)